File: Is Binary Tree Heap - GFG/is-binary-tree-heap.py
#User Template for python3

'''
class Node:
    def __init__(self,val):
        self.data = val
        self.left = None
        self.right = None
'''

class Solution:
    #Your Function Should return True/False
    def isHeap(self, root):
        #Code Here
        if not root: return True
        queue = [(root, 1)]
        level = 0
        while queue:
            for i, v in enumerate(queue):
                if v[1] != i+1: return False
            new_queue = []
            for v, i in queue:
                if v.left:
                    if v.left.data > v.data: return False
                    new_queue.append((v.left, 2 * i - 1))
                if v.right:
                    if v.right.data > v.data: return False
                    new_queue.append((v.right, 2 * i))
            if new_queue and len(queue) != (1 << level): return False
            level += 1
            queue = new_queue
        return True
#{ 
 # Driver Code Starts
#Initial Template for Python 3

#Contributed by Susanta Mukherjee
import sys
sys.setrecursionlimit(10**6)
from collections import deque
# Tree Node
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None

    
# Function to Build Tree   
def buildTree(s):
    #Corner Case
    if(len(s)==0 or s[0]=="N"):           
        return None
        
    # Creating list of strings from input 
    # string after spliting by space
    ip=list(map(str,s.split()))
    
    # Create the root of the tree
    root=Node(int(ip[0]))                     
    size=0
    q=deque()
    
    # Push the root to the queue
    q.append(root)                            
    size=size+1 
    
    # Starting from the second element
    i=1                                       
    while(size>0 and i<len(ip)):
        # Get and remove the front of the queue
        currNode=q[0]
        q.popleft()
        size=size-1
        
        # Get the current node's value from the string
        currVal=ip[i]
        
        # If the left child is not null
        if(currVal!="N"):
            
            # Create the left child for the current node
            currNode.left=Node(int(currVal))
            
            # Push it to the queue
            q.append(currNode.left)
            size=size+1
        # For the right child
        i=i+1
        if(i>=len(ip)):
            break
        currVal=ip[i]
        
        # If the right child is not null
        if(currVal!="N"):
            
            # Create the right child for the current node
            currNode.right=Node(int(currVal))
            
            # Push it to the queue
            q.append(currNode.right)
            size=size+1
        i=i+1
    return root
    


if __name__=="__main__":
    t=int(input())
    for _ in range(0,t):
        root=buildTree(input())
        ob = Solution()
        if ob.isHeap(root):
            print(1)
        else:
            print(0)
        
        

# } Driver Code Ends

File: Is Binary Tree Heap - GFG/is-binary-tree-heap.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// Tree Node
struct Node {
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};

// Function to Build Tree
Node *buildTree(string str) {
    // Corner Case
    if (str.length() == 0 || str[0] == 'N') return NULL;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str;) ip.push_back(str);

    // Create the root of the tree
    Node *root = new Node(stoi(ip[0]));

    // Push the root to the queue
    queue<Node *> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size()) {

        // Get and remove the front of the queue
        Node *currNode = queue.front();
        queue.pop();

        // Get the current Node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N") {

            // Create the left child for the current Node
            currNode->left = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size()) break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N") {

            // Create the right child for the current Node
            currNode->right = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}


// } Driver Code Ends
// User Function template for C++

// Structure of node
/*struct Node {
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};*/

class Solution {
  public:
/*t countNode(struct Node* tree){
        if(tree==NULL) return 0;
        int ans=1+countNode(tree->left)+countNode(tree->right);
        return ans;
    }
    bool isBST(struct Node* tree,int index,int countnode){
        if(tree==NULL) return true;
        int left=2*index+1;
        int right=2*index+2;
        if(index>=countnode) return true;
        else{
            left=isBST(tree,left,countnode);
            right=isBST(tree,right,countnode);
            return (left && right);
        }
    }
    vector<int> bfs(struct Node* node){
        vector<int> ans;
      if(node==NULL) return ans;
     
      queue<Node*> q;
      q.push(node);
      while(!q.empty()){
          Node* n=q.front();
          ans.push_back(n->data);
          q.pop();
          
              if(n->left!=NULL) q.push(n->left);
              if(n->right!=NULL) q.push(n->right);
              
          
      }
      return ans;
    }
    bool isHeap(struct Node* tree) {
        // code here
      vector<int> a=bfs(tree);
     int i=0;
     int countnode=countNode(tree);
     int left=2*i+1;
     int right=2*i+2;
     for(int i=0;i<a.size();i++){
         if(!(a[i]>a[left]) || !(a[i]>a[right])|| !isBST(tree,0,countnode))return false;
     }
     return true;
    }*/
    int getTotalNodes(Node* root){
    if(root==NULL) return 0;
    return 1 + getTotalNodes(root->left) + getTotalNodes(root->right);
}
bool isHeap(Node* root, int rootIndex, int totalNodes, int bound){
    if(root==NULL) return true;
    if(root->data>bound) return false;
    if(rootIndex>=totalNodes) return false;
    return isHeap(root->left, 2*rootIndex+1, totalNodes, root->data) && isHeap(root->right, 2*rootIndex+2, totalNodes, root->data);
}
bool isHeap(struct Node* tree) {
    int totalNodes = getTotalNodes(tree);
    return isHeap(tree, 0, totalNodes, tree->data);
}

};

//{ Driver Code Starts.

int main() {
    int tc;
    scanf("%d ", &tc);
    while (tc--) {
        string treeString;
        getline(cin, treeString);
        Solution ob;
        Node *root = buildTree(treeString);
        if (ob.isHeap(root))
            cout << 1 << endl;
        else
            cout << 0 << endl;
    }

    return 0;
}
// } Driver Code Ends